src/models/room.js

```diff
@@ -1277,19 +1291,37 @@ Room.prototype.updatePendingEvent = function(event, newStatus, newEventId) {
     } else if (newStatus == EventStatus.CANCELLED) {
         // remove it from the pending event list, or the timeline.
         if (this._pendingEventList) {
-            utils.removeElement(
-                this._pendingEventList,
-                function(ev) {
-                    return ev.getId() == oldEventId;
-                }, false,
-            );
+            const idx = this._pendingEventList.findIndex(ev => ev.getId() === oldEventId);
+            if (idx !== -1) {
+                const [removedEvent] = this._pendingEventList.splice(idx, 1);
+                if (removedEvent.getType() === "m.room.redaction") {
+                    this._revertRedactionLocalEcho(removedEvent);
+                }
+            }
         }
         this.removeEvent(oldEventId);
     }
 
     this.emit("Room.localEchoUpdated", event, this, oldEventId, oldStatus);
 };
 
+Room.prototype._revertRedactionLocalEcho = function(redactionEvent) {
+    const redactId = redactionEvent.event.redacts;
+    if (!redactId) {
+        return;
+    }
+    const redactedEvent = this.getUnfilteredTimelineSet()
+        .findEventById(redactId);
+    if (redactedEvent) {
+        redactedEvent.unmarkLocallyRedacted();
+        // re-render after undoing redaction
+        this.emit("Room.redaction", redactionEvent, this);
```

Comment:
Would a different event be better, in case listeners want to take a separate action for undoing reactions? Or is there a way to distinguish the two from the args?

Reply:
Initially I added a boolean flag as the third parameter, but nothing would use it. Only the `TimelinePanel` listens for `Room.redaction" and it does a `forceUpdate`, which is also what we do now on "unredaction". Seemed a bit artificial to add an event that calls the same event handler there, or duplicates it?

Reply:
Well, since the JS SDK has other consumer than just the React SDK, I wonder if others would want handle them differently...? (I have no idea.) I guess I am just thinking that if I were using this SDK in my app, I might be surprised by the same event be reused for the opposite action. Up to you though, maybe we just wait and see if hear it's a concern...?

Reply:
That is a fair point, I'll add a `Room.redactionCancelled` event then.